Feature/leases

jessicayu626 opened the pull request:
Implemented renew_lease, recover_lease and complete with some tests on renew_lease and recover_lease.

Commit:
Add native file systems tests.

Commit:
pull

Commit:
Implement renewlease, recoverlease and complete.

Commit:
fix client path errors.

Commit:
add leasetest.

Commit:
Add lease tests.

Antzen commented on zookeeper/source/zk_client_common.cc:
```diff
@@ -28,6 +28,8 @@ const char ZkClientCommon::STATS[] = "/stats";
 const char ZkClientCommon::BLOCK_LOCATIONS[] = "/block_locations/";
 const char ZkClientCommon::BLOCKS[] = "/blocks";
+const char ZkClientCommon::LEASES[] = "/leases";
+const char ZkClientCommon::CLIENTS[] = "/clients";
```
Hm, did we agree that "/clients" should be a subdirectory of the metadata? (There is no "/clients" in the "RDFS ZooKeeper Metadata structure" Google doc)

If it is part of the metadata, how should it be located/defined in our current metadata scheme?

jessicayu626 replied:
Yes. It is a subdirectory.

LINKIWI committed:
Add WebRDFS HTTP server library to build (#141)

* Add Python-based linter virtualenv to .gitignore

* Simple-Web-Server in utility

* Rename library hpp -> h, cpp -> cc and remove examples from CMakeLists.txt

LINKIWI committed:
Add WebRDFS HTTP server library to build (#141)

* Add Python-based linter virtualenv to .gitignore

* Simple-Web-Server in utility

* Rename library hpp -> h, cpp -> cc and remove examples from CMakeLists.txt

vnazz committed:
Feature/webrdfs web request translator (#148)

* Add basic structure for webRequestTranslator and tests

* Add read, status and listing to webRequestTranslator

* add http responses for create, mkdir, mv delete

* Fix build issue on travis

* Hopefully fix permission issue in translator

* Jk this should fix permissions

* Add newlines to fix lint

* Add WebRDFS HTTP server library to build (#141)

* Add Python-based linter virtualenv to .gitignore

* Simple-Web-Server in utility

* Rename library hpp -> h, cpp -> cc and remove examples from CMakeLists.txt

* Change directory name and add more info to mv, delete, mkdir handlers

* Add basic structure for webRequestTranslator and tests

* Add read, status and listing to webRequestTranslator

* add http responses for create, mkdir, mv delete

* Fix build issue on travis

* Hopefully fix permission issue in translator

* Jk this should fix permissions

* Add newlines to fix lint

* Change directory name and add more info to mv, delete, mkdir handlers

* Replace underscores

vnazz committed:
Feature/webrdfs web request translator (#148)

* Add basic structure for webRequestTranslator and tests

* Add read, status and listing to webRequestTranslator

* add http responses for create, mkdir, mv delete

* Fix build issue on travis

* Hopefully fix permission issue in translator

* Jk this should fix permissions

* Add newlines to fix lint

* Add WebRDFS HTTP server library to build (#141)

* Add Python-based linter virtualenv to .gitignore

* Simple-Web-Server in utility

* Rename library hpp -> h, cpp -> cc and remove examples from CMakeLists.txt

* Change directory name and add more info to mv, delete, mkdir handlers

* Add basic structure for webRequestTranslator and tests

* Add read, status and listing to webRequestTranslator

* add http responses for create, mkdir, mv delete

* Fix build issue on travis

* Hopefully fix permission issue in translator

* Jk this should fix permissions

* Add newlines to fix lint

* Change directory name and add more info to mv, delete, mkdir handlers

* Replace underscores

Commit:
Fix lint errors.

Commit:
Fix build error

Antzen commented on zookeeper/source/zk_nn_client.cc:
```diff
@@ -273,6 +273,83 @@ char ZkNnClient::get_node_policy() {
+  }
+}
+
+void ZkNnClient::recover_lease(RecoverLeaseRequestProto &req, RecoverLeaseResponseProto &res) {
```
A doc-string that mentions what this method may be used for would be helpful here. Just want to clarify - is the plan here to have heartbeats check active leases by calling this, or is the client responsible for recovering leases when done with their write?

Antzen commented on zookeeper/source/zk_nn_client.cc:
```diff
@@ -291,6 +368,23 @@ void ZkNnClient::file_znode_struct_to_vec(FileZNode *znode_data,
+  std::uint8_t *buffer = &data[0];
+  memcpy(&znode_data, buffer, sizeof(znode_data));
+}
+
```
👍

Commit:
Fix more lint errors.

Commit:
Two more lint errors.

Antzen commented on test/native-filesystem/native_fs_tests.cc:
```diff
@@ -112,13 +115,85 @@ TEST_F(NativeFSTest, CanCoalesce) {
+    }
+    ASSERT_FALSE(filesystem.writeBlock(
+      (DISK_SIZE - RESERVED_SIZE) / MIN_BLOCK_SIZE,
+                                       smallerThanMinBlk_a));
```
👍 
Yeah, so following up our discussion these past few weeks - might be good to have a reserve block implementation if time allows for it.

Antzen commented on zookeeper/include/zk_client_common.h:
```diff
@@ -36,6 +36,8 @@ class ZkClientCommon {
   static const char BLOCK_LOCATIONS[];
   static const char BLOCKS[];
+  static const char LEASES[];
+  static const char CLIENTS[];
```
See comment below about "/leases" directory in our metadata.

Antzen commented on zookeeper/include/zk_nn_client.h:
```diff
@@ -45,6 +45,22 @@ typedef struct {
+ */
+typedef struct {
+  uint64_t timestamp;    // std::time()
+} ClientInfo;
```
I think these two structs for the leases and client statuses are fine for now, but as discussed in class, maybe it would be nice to have the filename in the ClientInfo and/or the timestamp in the LeaseInfo to avoid extra lookups. I'll mark this as something to revisit later if needed.

Antzen commented on zookeeper/source/zk_nn_client.cc:
```diff
@@ -273,6 +273,94 @@ char ZkNnClient::get_node_policy() {
+    znode_data_to_vec(&clientInfo, data);
+
+    if (!exists) {
+      if (!zk->create(ClientZookeeperPath(client_name),
```
Hm, so this is code for constructing a client status if it doesn't already exist? Maybe we should modularize this out since this is code we will use for constructing a client status (as opposed to renewing it).

LINKIWI committed:
Update Zookeeper mirror to Rice server (#167)

Antzen commented on zookeeper/source/zk_nn_client.cc:
```diff
@@ -721,6 +854,44 @@ void ZkNnClient::complete(CompleteRequestProto& req,
+      return;
+    }
+    if (exists) {
+      if (!zk->delete_node(LeaseZookeeperPath(src) + '/'
```
Is this repeated code from the RecoverLease method? Wouldn't it make sense to modularize out a "delete lease" method? (To be used by both the RecoverLease method and this method here)

Antzen commented on zookeeper/source/zk_nn_client.cc:
```diff
@@ -1180,6 +1351,14 @@ bool ZkNnClient::sort_by_xmits(const std::vector<std::string> &unsorted_dn_ids,
+  std::string zkpath = ZookeeperFilePath(hadoopPath);
+  zkpath += LEASES;
+  return zkpath;
+}
```
Okay, looks good (as we discussed); see comment above about metadata.

Antzen commented on zookeeper/source/zk_nn_client.cc:
```diff
@@ -1552,6 +1733,50 @@ bool ZkNnClient::rename_ops_for_file(const std::string &src,
+    << "' when renaming.";
+    return false;
+  }
+  ops.push_back(zk->build_create_op(dst_znode_leases, lease_data));
```
Not sure if it's worth doing the same thing here, but we should probably modularize out this method of constructing leases.

Antzen reviewed (commented):
Looks good. Just two general things of note (some details in comments):
1. We clarified the metadata discrepancy in person, so that should be fine now. I left my comments in for the record (google doc is now updated).
2. I think we can modularize some of the code to make it cleaner, if possible (see comments), since some of our code aren't unique to the protocol methods (e.g. constructing or deleting a lease)

Also, of course, we should figure out why the code isn't passing the tests.

LINKIWI committed:
Temporarily disable RW test (#168)

LINKIWI committed:
WebRDFS server process, placeholder HTTP handlers (#144)

* Simple-Web-Server in utility

* Add web-rdfs directory

* Main executable, handler registration

* Address lint errors

* Sample handlers registered

* send proto to zk in webrdfs

* Rename library hpp -> h, cpp -> cc and remove examples from CMakeLists.txt

* Correctly reference header files

* Lint errors

LINKIWI committed:
WebRDFS server process, placeholder HTTP handlers (#144)

* Simple-Web-Server in utility

* Add web-rdfs directory

* Main executable, handler registration

* Address lint errors

* Sample handlers registered

* send proto to zk in webrdfs

* Rename library hpp -> h, cpp -> cc and remove examples from CMakeLists.txt

* Correctly reference header files

* Lint errors

Commit:
Fix lease test.

danielbalkum committed:
Sprint 1: Permissions on a file (#92)

* Add getUserName method in rpcServer and simple test

* added user permissions listing to FileZNode and initialized permissions
fields during file creation in ZK.

* Set owner HDFS call modifies permissions ACL in ZK; fixes #53

* Modify zk_nn to pass along client's username as identity to zk. Add bool indicating whether in secure mode in zk_nn. Check access before serving files.

* debug

* Fix username test so passes both locally and on travis and renamed function getUserName to getUsername

* added user permissions listing to FileZNode and initialized permissions
fields during file creation in ZK.

* Set owner HDFS call modifies permissions ACL in ZK; fixes #53

* Fix merge issue on .ci/test/main.sh

* adding zookeeper metadata for keeping admin identity (#56)
adding logic to process change of file permissions (#55)

* Fix typo in name node main

* Fix segfault

* Statically-sized char arrays in ACL

* Fix lint errors

* Fix lint issues

* Order of header lint error fixed

* Fix username test lint issues

* All remaining lint errors addressed

* merging permissions changes

* lint errors after merge

* add docs for zk_nn_client.h

* fix on logic to determin admin

* fix lint

* Add getUserName method in rpcServer and simple test

* added user permissions listing to FileZNode and initialized permissions
fields during file creation in ZK.

* added user permissions listing to FileZNode and initialized permissions
fields during file creation in ZK.

* Set owner HDFS call modifies permissions ACL in ZK; fixes #53

* Set owner HDFS call modifies permissions ACL in ZK; fixes #53

* Modify zk_nn to pass along client's username as identity to zk. Add bool indicating whether in secure mode in zk_nn. Check access before serving files.

* debug

* Fix username test so passes both locally and on travis and renamed function getUserName to getUsername

* Fix merge issue on .ci/test/main.sh

* adding zookeeper metadata for keeping admin identity (#56)
adding logic to process change of file permissions (#55)

* Fix typo in name node main

* Fix segfault

* Statically-sized char arrays in ACL

* Fix lint errors

* Add getUserName method in rpcServer and simple test

* Fix lint issues

* Order of header lint error fixed

* Fix username test lint issues

* All remaining lint errors addressed

* merging permissions changes

* lint errors after merge

* add docs for zk_nn_client.h

* fix on logic to determin admin

* fix lint

* fixing error made while fixing rebase conflicts

heliumj committed:
generate cert (#171)

(cherry picked from commit 5109faaad3e7f61d057c4292ce5360fb4e089bef)

Commit:
Merge branch 'develop' of https://github.com/comp413-2017/RDFS into feature/leases

Commit:
Fix commit error.

Commit:
Fix build errors.

Commit:
Fix build error

Commit:
Debug test.

Commit:
Fix lint errors.

Commit:
Fix lint error.

Commit:
Add lease test executable.

Antzen reviewed (approved):
Looks good to me. Thanks for making the fixes.
